### PR TITLE
shared_ptr-ize the llvm::Module*.

### PR DESCRIPTION
**Pull request by vgvassilev**
This is in prepare for the upcoming llvm upgrade. The future orc jit compile
layer needs a std::shared_ptr<llvm::Module>. The current design passes a
llvm::Module* around and any conversions to a shared_ptr cause the
destruction of the llvm::Module which is a long-lived object in cling.

### CODE REVIEW

File: interpreter/cling/.clang-format [hunk 1]
```diff
@@ -1,6 +1,7 @@
 
 BasedOnStyle: LLVM
 IndentWidth: 2
+NamespaceIndentation: All
```
Comment (Axel-Naumann) [inline interpreter/cling/.clang-format]: Can you make this a separate commit?

File: interpreter/cling/include/cling/Interpreter/ClangInternalState.h [hunk 2]
```diff
@@ -44,16 +44,16 @@ namespace cling {
     const clang::ASTContext& m_ASTContext;
     const clang::Preprocessor& m_Preprocessor;
     clang::CodeGenerator* m_CodeGen;
-    const llvm::Module* m_Module;
+    std::shared_ptr<const llvm::Module> m_Module;
```
Comment (Axel-Naumann) [inline interpreter/cling/include/cling/Interpreter/ClangInternalState.h]: Why would the `ClangInternalState` need to own? This should be a `weak_ptr`.

File: interpreter/cling/include/cling/Interpreter/ClangInternalState.h [hunk 3]
```diff
@@ -44,16 +44,16 @@ namespace cling {
...
     const std::string m_DiffCommand;
     const std::string m_Name;
     ///\brief Takes the ownership after compare was made.
     ///
     std::unique_ptr<ClangInternalState> m_DiffPair;
   public:
     ClangInternalState(const clang::ASTContext& AC, const clang::Preprocessor&,
-                       const llvm::Module* M, clang::CodeGenerator* CG,
-                       const std::string& name);
+                       std::shared_ptr<const llvm::Module> M,
```
Comment (Axel-Naumann) [inline interpreter/cling/include/cling/Interpreter/ClangInternalState.h]: It's cheaper to take the ownership-counting (or even `weak_ptr`) by `const&`.

File: interpreter/cling/include/cling/Interpreter/ClangInternalState.h [hunk 4]
```diff
@@ -83,7 +83,7 @@ namespace cling {
 
     ///\brief Return the llvm::Module this state is bound too.
     ///
-    const llvm::Module* getModule() const { return m_Module; }
+    std::shared_ptr<const llvm::Module> getModule() const { return m_Module; }
```
Comment (Axel-Naumann) [inline interpreter/cling/include/cling/Interpreter/ClangInternalState.h]: Returning a `weak_ptr` (or `shared_ptr`) by `const&` is cheaper.

File: interpreter/cling/include/cling/Interpreter/Transaction.h [hunk 5]
```diff
@@ -463,8 +463,8 @@ namespace cling {
         m_NestedTransactions->clear();
     }
 
-    llvm::Module* getModule() const { return m_Module.get(); }
-    void setModule(std::unique_ptr<llvm::Module> M) { m_Module.swap(M); }
+    std::shared_ptr<llvm::Module> getModule() const { return m_Module; }
```
Comment (Axel-Naumann) [inline interpreter/cling/include/cling/Interpreter/Transaction.h]: `const&`, if possible.

File: interpreter/cling/lib/Interpreter/ClangInternalState.cpp [hunk 6]
```diff
@@ -36,7 +36,7 @@ namespace cling {
 
   ClangInternalState::ClangInternalState(const ASTContext& AC,
                                          const Preprocessor& PP,
-                                         const llvm::Module* M,
+                                         std::shared_ptr<const llvm::Module> M,
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/ClangInternalState.cpp]: Should probably take a `const weak_ptr&` directly a `const shared_ptr&`

File: interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp [hunk 7]
```diff
@@ -124,8 +124,8 @@ void IncrementalExecutor::shuttingDown() {
   }
 }
 
-void IncrementalExecutor::AddAtExitFunc(void (*func) (void*), void* arg,
-                                        llvm::Module* M) {
+void IncrementalExecutor::AddAtExitFunc(void (*func)(void*), void* arg,
+                                        std::shared_ptr<llvm::Module> M) {
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp]: Should this entry in the exit functions *own* the module? If not `const weak_ptr&`

File: interpreter/cling/lib/Interpreter/IncrementalExecutor.h [hunk 8]
```diff
@@ -116,7 +116,8 @@ namespace cling {
     ///\brief Function registered via __cxa_atexit, atexit, or one of
     /// it's C++ overloads that should be run when a module is unloaded.
     ///
-    typedef utils::OrderedMap<llvm::Module*, std::vector<CXAAtExitElement>>
+    typedef utils::OrderedMap<std::shared_ptr<llvm::Module>,
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/IncrementalExecutor.h]: Should probably be a non-owning `weak_ptr`.

File: interpreter/cling/lib/Interpreter/IncrementalExecutor.h [hunk 9]
```diff
@@ -171,8 +172,10 @@ namespace cling {
     }
 
     ///\brief Unload a set of JIT symbols.
-    bool unloadFromJIT(llvm::Module* M, Transaction::ExeUnloadHandle H) {
-      auto iMod = std::find(m_ModulesToJIT.begin(), m_ModulesToJIT.end(), M);
+    bool unloadFromJIT(std::shared_ptr<llvm::Module> M,
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/IncrementalExecutor.h]: Here, ownership might make sense (during the lifetime of the unload operation). But then this should take a `const shared_ptr&`.

File: interpreter/cling/lib/Interpreter/Interpreter.cpp [hunk 10]
```diff
@@ -86,7 +86,7 @@ namespace {
     return cling::Interpreter::kExeSuccess;
   }
 
-  static bool isPracticallyEmptyModule(const llvm::Module* M) {
+  static bool isPracticallyEmptyModule(std::shared_ptr<const llvm::Module> M) {
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/Interpreter.cpp]: Should just take a `Module*`; I don't see that ownership should be involved here?

File: interpreter/cling/lib/Interpreter/Interpreter.cpp [hunk 11]
```diff
@@ -125,11 +125,10 @@ namespace cling {
...
       PushTransactionRAII pushedT(i);
 
-      m_State.reset(new ClangInternalState(CI.getASTContext(),
-                                           CI.getPreprocessor(),
-                                           CG ? CG->GetModule() : 0,
-                                           CG,
-                                           "aName"));
+      m_State.reset(new ClangInternalState(
+          CI.getASTContext(), CI.getPreprocessor(),
+          CG ? std::shared_ptr<llvm::Module>(CG->GetModule()) : nullptr, CG,
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/Interpreter.cpp]: This looks strange - should the `ClangInternalState` take ownership of the `Module` produced by `CodeGen`? I'd have bet that the `Transaction` should take ownership, and the `ClangInternalState` gets "observer status" by getting passed a `weak_ptr`.

File: interpreter/cling/lib/Interpreter/TransactionUnloader.cpp [hunk 12]
```diff
@@ -147,7 +147,7 @@ namespace cling {
     return cling::UnloadDecl(m_Sema, m_CodeGen, D);
   }
 
-  bool TransactionUnloader::unloadModule(llvm::Module* M) {
+  bool TransactionUnloader::unloadModule(std::shared_ptr<llvm::Module> M) {
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/TransactionUnloader.cpp]: `const shared_ptr&`

File: interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp (hunk 7 above)
Comment (Axel-Naumann) [reply interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp]: Make it a `const shared_ptr&`; I agree it's a good sanity check that the `Module` is not owned by the AtExitFuncs anymore.

Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp]: I don't find this pattern in the codebase of llvm. I assume nowadays compilers can use move semantics.

Comment (pcanal) [reply interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp]: Does AddAtExitFunc has any 'possibility' of recording the pointer or is it just using it during its execution?  If it is the later, you might consider just passing the raw pointer (or const shared_ptr&).

Comment (Axel-Naumann) [reply interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp]: Vassil wants the ExitFunc to keep ref counts. See https://stackoverflow.com/questions/37610494/passing-const-shared-ptrt-versus-just-shared-ptrt-as-parameter why that should be fixed in the codebase of llvm :-)

File: interpreter/cling/include/cling/Interpreter/Transaction.h (hunk 5 above)
Comment (vgvassilev) [reply interpreter/cling/include/cling/Interpreter/Transaction.h]: When the `shared_ptr` is used as a return type, RVO and automatic moves make it no different.

File: interpreter/cling/lib/Interpreter/TransactionUnloader.cpp (hunk 12 above)
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/TransactionUnloader.cpp]: Done.

File: interpreter/cling/lib/Interpreter/IncrementalExecutor.h (hunk 9 above)
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/IncrementalExecutor.h]: Done.

File: interpreter/cling/lib/Interpreter/IncrementalExecutor.h (hunk 8 above)
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/IncrementalExecutor.h]: Added a FIXME.

File: interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp (hunk 7 above)
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp]: Done.
